### PR TITLE
Bug in MultidomainFunction python wrapper

### LINKED ISSUE #21102: Bug in MultidomainFunction python wrapper
### Expected behaviour

This code should run without error:
```
fun = MultiDomainFunction(Gaussian(PeakCentre=1, Sigma=0.1), Gaussian(PeakCentre=1, Sigma=0.2),  Global=['Height'])

```

### Actual behaviour

Exception raised:
```
AttributeError: 'str' object has no attribute 'fun'
  at line 1 in 'New script'
  caused by line 656 in 'C:\Users\Roman\Work\mantid\build\bin\Release\mantid\fitfunctions.py'
  caused by line 621 in 'C:\Users\Roman\Work\mantid\build\bin\Release\mantid\fitfunctions.py'
  caused by line 373 in 'C:\Users\Roman\Work\mantid\build\bin\Release\mantid\fitfunctions.py'
```

### Steps to reproduce the behavior

Run the code above.

### Platforms affected

All

### PR DESCRIPTION
**Pull request by mantid-roman**
Generic function that creates fit function wrappers for composite functions uses correct arguments now.

**To test:**

* Use the script form the issue description. Print out the created fit function to see that the global parameters are tied to each other.
* Run the code from @AnthonyLim23's comment. Gradient 1 and Gradient 2 must be both equal to 3.5

Fixes #21102

**Release Notes** 

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [x] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/test/python/mantid/FitFunctionsTest.py [hunk 1]
```diff
@@ -491,12 +491,17 @@ def test_iteration(self):
...
     def test_productfunction_creation(self):
         g0 = FunctionWrapper( "Gaussian", Height=7.5, Sigma=1.2, PeakCentre=10)
         g1 = FunctionWrapper( "Gaussian", Height=8.5, Sigma=1.2, PeakCentre=11)
         testhelpers.assertRaisesNothing(self, ProductFunctionWrapper, g0, g1)
 
+    def test_productfunction_creation_1(self):
+        g0 = Gaussian(Height=7.5, Sigma=1.2, PeakCentre=10)
+        g1 = Gaussian(Height=8.5, Sigma=1.2, PeakCentre=11)
+        testhelpers.assertRaisesNothing(self, ProductFunction, g0, g1)
+
```
Comment (KarlPalmen) [inline Framework/PythonInterface/test/python/mantid/FitFunctionsTest.py]: I'd use a more informative test name suffix than ``..._1``, such as ``..._by_name``. There is one more example of this.

**Comment by AnthonyLim23** [comment]
the code looks fine and it doesn't crash. However, the ties do not seem to work correctly in MultiDomainFunction (and the wrapper). The following is based on the last example http://docs.mantidproject.org/nightly/algorithms/Fit-v1.html

```
import math
import numpy as np

# create data
xData=np.linspace(start=0,stop=10,num=22)
yData=[]
for x in xData:
    yData.append(2.0*x+10.)
yData2=[]
for x in xData:
    yData2.append(5.0*x+7.)
# create workspaces
input = CreateWorkspace(xData,yData)
input2 = CreateWorkspace(xData,yData2)
# create function
myFunc = LinearBackground()
myFunc2 = LinearBackground()
multiFunc=MultiDomainFunction(myFunc,myFunc2,NumDeriv=1,Global=["A1"])
#multiFunc.tie({'f0.A1' : 'f1.A1'})
# do fit
fit_output = Fit(Function=multiFunc, InputWorkspace=input, WorkspaceIndex=0, \
                 InputWorkspace_1=input2, WorkspaceIndex_1=0, \
                 StartX = 0.1, EndX=9.5, StartX_1 = 0.1, EndX_1=9.5,Output='fit') 
paramTable = fit_output.OutputParameters  # table containing the optimal fit parameters # print results print('Gradients (shared):') print("Gradient 1: {0:.2f}".format(paramTable.column(1)[3]))
print("Gradient 1: {0:.2f}".format(paramTable.column(1)[3]))
print("Gradient 2: {0:.2f}".format(paramTable.column(1)[1]))
print('offsets:')
print("Constant 1: {0:.2f}".format(paramTable.column(1)[0]))
print("Constant 2: {0:.2f}".format(paramTable.column(1)[2]))
```

**Review by KarlPalmen — CHANGES_REQUESTED** [review]
Change fixes the problem and the result function looks OK.  I'm not happy with the names of the new unit tests and want something more informative, such as the suggestion I made.

If ties don't work in MultiDomainFunction rather than just the wrapper, then this is a different issue, which needs to be found or created.

**Review by KarlPalmen — APPROVED** [review]

**Review by AnthonyLim23 — CHANGES_REQUESTED** [review]
the code looks good and it all works now. However, the doc should be updated (it currently has global instead of Global)

**Comment by mantid-roman** [comment]
I would prefer `global`. I'll better change the code.

**Comment by AnthonyLim23** [comment]
as long as it is consistent I do not mind

**Comment by mantid-roman** [comment]
Oops. `global` is a reserved word. :(

**Review by AnthonyLim23 — APPROVED** [review]

**Comment by KarlPalmen** [comment]
I recall that is the reason chose ``Global`` rather than ``global`` (reserved word in Python).

File: Framework/PythonInterface/mantid/fitfunctions.py [hunk 2]
```diff
@@ -1,377 +1,386 @@
...
+        an Ifunction object to wrap.
+        :param **kwargs: standard argument for __init__ function
+        """
+        if not isinstance(name, str):
+            self.fun = name
+        else:
+            self.fun = FunctionFactory.createFunction(name)
+        self.init_paramgeters_and_attributes(**kwargs)
+
+    def init_paramgeters_and_attributes(self, **kwargs):
```
Comment (jmborr) [inline Framework/PythonInterface/mantid/fitfunctions.py]: is `paramgeters` the intended word in method `init_paramgeters_and_attributes` ?

**Pull request merged**